File: PresentControllerDemo/PresentControllerDemo/ViewController.m
//
//  ViewController.m
//  PresentControllerDemo
//
//  Created by alexiuce  on 2017/8/7.
//  Copyright © 2017年 alexiuce . All rights reserved.
//

#import "ViewController.h"
#import "TwoController.h"
#import "PresentAnimator.h"


@interface ViewController ()



@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view.
}



- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

- (IBAction)presentController:(NSButton *)sender {


     TwoController *twoVc  = [[TwoController alloc]init];
    
   /**  1. 以 sheet 方式(共享当前的window),弹出后,当前windows close 不可用 */
//    [self presentViewControllerAsSheet:twoVc];
    
    /** 2. 以modal 窗口的方式弹出新控制器(有独立的window)  */
//    [self presentViewControllerAsModalWindow:twoVc];
    
    /** 3. 自定义方式弹控制器, 需提供一个实现<NSViewControllerPresentationAnimator> 协议的类,完成动画 */
    
    PresentAnimator *modalAnimator = [[PresentAnimator alloc]init];
    [self presentViewController:twoVc animator:modalAnimator];
    
    /** 4. 以 popover 方式弹出新的控制器
     asPopoverRelativeToRect : 弹出popover时,附着的矩形区域(即被popover剪头指向的那一块矩形区域)
     ofView: 矩形区域( asPopoverRelativeToRect)所属的view
     preferredEdge:矩形区域( asPopoverRelativeToRect)的参考边界:
                     NSRectEdgeMinX:(在矩形区域左边显示popover)
                     NSRectEdgeMaxX:(在矩形区域右边显示popover)
                     NSRectEdgeMinY:(在矩形区域顶部显示popover)
                     NSRectEdgeMaxY:(在矩形区域底部显示popover)
     behavior : popover的行为模式
                NSPopoverBehaviorApplicationDefined :点击popover 视图以外的操作不会自动关闭(不支持ESC键盘关闭)
                NSPopoverBehaviorTransient :  点击popover 视图以外的操作会自动关闭(支持ESC键盘关闭)
                NSPopoverBehaviorSemitransient: : 点击popover 视图以外的操作不会自动关闭(支持ESC键盘关闭)
     
     */
    
//    [self presentViewController:twoVc asPopoverRelativeToRect:sender.frame ofView:self.view preferredEdge:NSRectEdgeMinY behavior:NSPopoverBehaviorApplicationDefined];
    
}

@end

File: NSTaskDemo/NSTaskDemo/ViewController.swift
//
//  ViewController.swift
//  NSTaskDemo
//
//  Created by alexiuce  on 2017/7/27.
//  Copyright © 2017年 alexiuce . All rights reserved.
//

import Cocoa

let kSelectedFilePath = "userSelectedPath"


class ViewController: NSViewController {

    @IBOutlet weak var repoPath: NSTextField!           // git 仓库path
    @IBOutlet weak var savePath: NSTextField!           // 本地保存路径
    
    var isLoadingRepo  = false                                 // 记录是否正在加载中..
    
    var outputPipe = Pipe()
    var task : Process?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }

    @IBAction func selectPath(_ sender: NSButton) {
        // 1. 创建打开文档面板对象
        let openPanel = NSOpenPanel()
        // 2. 设置确认按钮文字
        openPanel.prompt = "Select"
        // 3. 设置禁止选择文件
        openPanel.canChooseFiles = true
        // 4. 设置可以选择目录
        openPanel.canChooseDirectories = true
        // 5. 弹出面板框
        openPanel.beginSheetModal(for: self.view.window!) { (result) in
            // 6. 选择确认按钮
            if result == NSModalResponseOK {
                // 7. 获取选择的路径
                self.savePath.stringValue = (openPanel.directoryURL?.path)!
                UserDefaults.standard.setValue(openPanel.url?.path, forKey: kSelectedFilePath)
                UserDefaults.standard.synchronize()
            }
            // 8. 恢复按钮状态
            sender.state = NSOffState
        }
    }
    @IBAction func startPull(_ sender: NSButton) {
        guard  let executePath = UserDefaults.standard.value(forKey: kSelectedFilePath) as? String else {
            print("no selected path")
            return
        }
        guard self.repoPath.stringValue.characters.count != 0 else {
            return
        }
        if isLoadingRepo {return}   // 如果正在执行,则返回
        isLoadingRepo = true   // 设置正在执行标记
        task = Process()     // 创建NSTask对象
        
        task?.launchPath = "/bin/bash"    // 执行路径(这里是需要执行命令的绝对路径)
        task?.arguments = ["-c","cd \(executePath); git clone \(self.repoPath.stringValue)"]
        
        // 获取输出
        outputPipe = Pipe()
        task?.standardOutput = outputPipe
       
        outputPipe.fileHandleForReading.waitForDataInBackgroundAndNotify()
        NotificationCenter.default.addObserver(forName: NSNotification.Name.NSFileHandleDataAvailable, object: outputPipe.fileHandleForReading, queue: nil) { (notification) in
            let outputData = self.outputPipe.fileHandleForReading.availableData
            guard let outputString = String(data: outputData, encoding: .utf8) , outputString != "" else{return}
            print(outputString);
        }
    
        task?.terminationHandler = { proce in              // 执行结束的闭包(回调)
            self.isLoadingRepo = false    // 恢复执行标记
            print("finished")
            NotificationCenter.default.removeObserver(self, name: NSNotification.Name.NSFileHandleDataAvailable, object: self.outputPipe.fileHandleForReading)
        }
        task?.launch()                // 开启执行
        print("end")
        // 获取运行结果
//        let resultData = outputPipe.fileHandleForReading.readDataToEndOfFile()
//        let text = String(data: resultData, encoding: String.Encoding.utf8)
//        print(text ?? "non")
//        task?.waitUntilExit()       // 阻塞直到执行完毕
        
    }
}

